Introduction to jsx (#7)
reactjs/de.reactjs.org: content/docs/introducing-jsx.md

---
id: introducing-jsx
title: Einführung in JSX
permalink: docs/introducing-jsx.html
prev: hello-world.html
next: rendering-elements.html
---

Betrachte die folgende Variablendeklaration:

```js
const element = <h1>Hallo Welt!</h1>;
```

Diese seltsame Tag-Schreibweise ist weder ein String noch HTML.

Das ist eine syntaktische Erweiterung zu JavaScript und heißt JSX. Wir empfehlen dies zusammen mit React zu benutzen um zu beschreiben wie das UI aussehen soll. JSX mag vielleicht an eine Template-Sprache erinnern, hat jedoch den vollen Funktionsumfang von JavaScript. 

JSX erzeugt React-Elemente "elements". Wie diese im DOM gerendered werden, behandeln wir im nächsten Kapitel [next section](/docs/rendering-elements.html). Lies weiter, um die Grundlagen über JSX zu lernen, die du zum Starten benötigst.

### Warum JSX? {#why-jsx}

> React lebt den Fakt, dass die Logik des Renderings und andere UI-Logiken grundsätzlich miteinander verbunden sind: Wie Events behandelt werden, wie sich der State über die Zeit verändert und wie Daten für die Darstellung vorbereitet werden.

Reviewer: ```suggestion
Die Tatsache in React ist, dass die Rendering und UI-Logik grundsätzlich miteinander verbunden ist: Wie Events behandelt werden, wie sich der State über die Zeit verändert und wie Daten für die Darstellung vorbereitet werden.
```

Author: ```suggestion
React verfolgt das Konzept, dass die Rendering- und UI-Logik grundsätzlich miteinander verbunden ist: Wie Events behandelt werden, wie sich der State über die Zeit verändert und wie Daten für die Darstellung vorbereitet werden.
```

Author: @ph1p what do you think about that version?

Reviewer: Sounds good to me :)

> Anstelle das *Technologien* künstlich separiert werden, indem Markup und Logik in getrennten Datein liegen, trennt React die Dateien [nach *Zuständigkeit*](https://en.wikipedia.org/wiki/Separation_of_concerns) auf und bindet die daraus resultierenden Einheiten in sogenannten "components" lose zusammen. Diese Components enthalten sowohl Markup als auch Logik. Zu den Components werden wir noch [in einem anderen Kapitel](/docs/components-and-props.html) kommen. Wenn Du dich immernoch unwohl dabei fühlst, Markup in JavaScript zu schreiben, dann schaue Dir doch am Besten [diesen Talk](https://www.youtube.com/watch?v=x7cQ3mrcKaY) an.

Reviewer: ```suggestion
Anstelle das *Technologien* künstlich separiert werden, indem Markup und Logik in getrennten Dateien liegt, trennt React die Dateien [nach *Zuständigkeit*](https://en.wikipedia.org/wiki/Separation_of_concerns) auf und bindet die daraus resultierenden Einheiten in sogenannten Komponenten ("components") lose zusammen. Diese Komponenten enthalten sowohl Markup als auch Logik. Zu den Komponenten werden wir noch [in einem anderen Kapitel](/docs/components-and-props.html) mehr erfahren. Wenn Du dich immer noch unwohl dabei fühlst, Markup in JavaScript zu schreiben, dann schaue Dir doch am Besten [diesen Talk](https://www.youtube.com/watch?v=x7cQ3mrcKaY) an.
```

Author: i would commit this suggestion, but i am not sure about "indem Markup und Logik in getrennten Dateien liegt" -> why liegt instead of liegen?

Reviewer: I've submitted a new suggestion @ChrisB9

[](codepen://introducing-jsx)

Aus gründen der Lesbarkeit haben wir JSX auf mehrere Zeilen verteilt. Dies ist prinzipiell nicht notwendig und wenn man es macht, empfehlen wir, Klammern drumherum zu setzen, um etwaige Fehler durch das [automatische Einfügen von Semikolons](http://stackoverflow.com/q/2846283) zu vermeiden.

### JSX ist auch ein Ausdruck {#jsx-is-an-expression-too}

> Nach der Kompilierung werden JSX Ausdrücke normale JavaScript-Funktionssaufrufe und werden zu Objekten evaluiert.

Reviewer: ```suggestion
Nach dem Kompilieren werden JSX Ausdrücke normale JavaScript-Funktionssaufrufe und als JavaScript-Objekte ausgewertet.
```

> Das bedeutet, dass JSX innerhalb von  `if`-Blöcken und `for`-Schleifen stehen, Variablen zugewiesen, als Argument übergeben oder Rückgabewert eine Funktion sein kann:

Reviewer: ```suggestion
Das bedeutet, dass JSX innerhalb von  `if`-Blöcken und `for`-Schleifen stehen, Variablen zugewiesen, als Argument übergeben oder Rückgabewert einer Funktion sein kann:
```

```js{3,5}
function getGreeting(user) {
  if (user) {
    return <h1>Hallo {formatName(user)}!</h1>;
  }
  return <h1>Hallo Fremder.</h1>;
}
```

> ### Mit JSX Attribute spezifizieren {#specifying-attributes-with-jsx}

Reviewer: ```suggestion
### Attribute mit JSX festlegen {#specifying-attributes-with-jsx}
```

Benutze Anführungszeichen um string-Literale als Attribute zu verwenden:

```js
const element = <div tabIndex="0"></div>;
```

Ebenfalls kannst du geschweifte Klammern verwenden um JavaScript-Ausdrücke in ein Attribut einzubinden:

```js
const element = <img src={user.avatarUrl}></img>;
```

Setze keine Anführungszeichen um geschweifte Klammern um JavaScript-Ausdrücke in ein Attribut einzubinden. Benutze entweder Anführungszeichen (für Strings) oder geschweifte Klammern (für Ausdrücke) aber nicht beides zusammen im selben Attribut.

>**Warnung:**
>
>Da JSX näher an JavaScript als an HTML ist, verwendet React DOM `camelCase` als Namenskonvention für Eigenschaften anstelle der für HTML typischen schreibweise.
>

> >`class` wird zum Beispiel zu [`className`](https://developer.mozilla.org/en-US/docs/Web/API/Element/className) in JSX, und `tabindex` wird zu [`tabIndex`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/tabIndex).

Reviewer: ```suggestion
>`class` wird in JSX zum Beispiel zu [`className`](https://developer.mozilla.org/en-US/docs/Web/API/Element/className), und `tabindex` zu [`tabIndex`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/tabIndex).
```

> ### Spezifizieren von Kind-Elementen mit JSX {#specifying-children-with-jsx}

Reviewer: ```suggestion
### Kind-Elemente mit JSX angeben {#specifying-children-with-jsx}
```

> ### Spezifizieren von Kind-Elementen mit JSX {#specifying-children-with-jsx}

Reviewer: ```suggestion
### Kind-Elemente in JSX angeben {#specifying-children-with-jsx}
```

> Ist ein Element leer, kannst du es wie in der XML-Notation mit `/>` schließen:

Reviewer: ```suggestion
Ist ein Element leer, kannst du es wie in XML mit `/>` schließen:
```

```js
const element = <img src={user.avatarUrl} />;
```

JSX-Tags können Kind-Elemente enthalten:

```js
const element = (
  <div>
    <h1>Hallo!</h1>
    <h2>Schön dich hier zu sehen.</h2>
  </div>
);
```

### JSX verhindert Injection-Angriffe {#jsx-prevents-injection-attacks}

> JSX kann sicher für Nutzereingaben verwendet werden:

Reviewer: ```suggestion
Es ist sicher, mit JSX Nutzereingaben zu tätigen:
```

Author: @ph1p i am pretty troubled with the "sicher", because of the two meanings of that word.. do you think it's safe to assume that the user understands "sicher" as in secure and not as in "sure"

Reviewer: You're right! I've updated my text.

```js
const title = response.potentiallyMaliciousInput;
// Das dagegen ist abgesichert:
const element = <h1>{title}</h1>;
```

> Standardmäßig [escaped](http://stackoverflow.com/questions/7381974/which-characters-need-to-be-escaped-on-html) React DOM jeden in JSX eingebetteten Wert vor dem Darstellen. Damit wird sichergestellt, dass niemals etwas in die Anwendung gelangt, dass nicht explizit so programmiert wurde. Alles wird zu einem String konvertiert und danach erst gerendered. Das hilft [XSS (cross-site-scripting)](https://en.wikipedia.org/wiki/Cross-site_scripting)-Attakten vorzubeugen.

Reviewer: ```suggestion
Standardmäßig [escaped](http://stackoverflow.com/questions/7381974/which-characters-need-to-be-escaped-on-html) React DOM jeden in JSX eingebetteten Wert vor dem Darstellen. Damit wird sichergestellt, dass niemals etwas in die Anwendung gelangt, dass nicht explizit so implementiert wurde. Alles wird zu einem String konvertiert und danach erst gerendert. Das hilft [XSS (cross-site-scripting)](https://en.wikipedia.org/wiki/Cross-site_scripting)-Attakten vorzubeugen.
```

const element = React.createElement(
  'h1',
  {className: 'greeting'},
  'Hallo Welt!'
);
```

> `React.createElement()` prüft erst die Eingabeparameter, um dich dabei zu unterstützen Bug-Frei zu programmieren aber im Prinzip erstellt es ein Objekt wie dies:

Reviewer: ```suggestion
`React.createElement()` prüft erst die Eingabeparameter, um dich dabei zu unterstützen bugfrei zu programmieren, aber im Prinzip erstellt es ein Objekt wie dieses:
```

```js
// Hinweis: Dies ist eine vereinfachte Struktur
const element = {
  type: 'h1',
  props: {
    className: 'greeting',
    children: 'Hallo Welt!'
  }
};
```

> Diese Objekte nennen sich "React elements". Stell sie dir als Beschreibung, was du auf dem Bildschirm sehen willst, vor. React liest diese Objekte und verwendet sie um den DOM zu erstellen und aktuell zu halten.

Reviewer: ```suggestion
Diese Objekte nennen sich "React elements". Stell sie dir als Beschreibung für das was du auf dem Bildschirm sehen willst, vor. React liest diese Objekte und verwendet sie um das DOM zu erstellen und es aktuell zu halten.
```

> Im nächsten Abschnitt gehen wir auf das rendering von React-Elementen in den DOM.

Reviewer: ```suggestion
Im nächsten Abschnitt gehen wir auf das Rendering von React-Elementen ins DOM ein.
```
Das "Document Object **Model**"

>**Tipp:**
>

> >Wir empfehlen die Verwendung der ["Babel" Sprachdefinition](http://babeljs.io/docs/editors) für den Editor deiner Wahl, sodass sowohl ES6 als auch JSX-Code vernünftig dargestellt werden kann. Diese Webseite nutzt das [Oceanic Next](https://labs.voronianski.com/oceanic-next-color-scheme/)-Farbschema, welched mit Babel kompatibel ist.

Reviewer: ```suggestion
>Wir empfehlen die Verwendung der ["Babel" Sprachdefinition](http://babeljs.io/docs/editors) für den Editor deiner Wahl, sodass sowohl ES6 als auch JSX-Code vernünftig dargestellt werden kann. Diese Webseite nutzt das [Oceanic Next](https://labs.voronianski.com/oceanic-next-color-scheme/)-Farbschema, welches mit Babel kompatibel ist.
```
